symbolic-base/src/ZkFold/Base/Algebra/Polynomials/Univariate.hs

```diff
@@ -67,28 +44,88 @@ infixl 6 .+, +.
 
 -------------------------------- Arbitrary degree polynomials --------------------------------
 
--- TODO (Issue #17): hide constructor
-newtype Poly c = P (V.Vector c)
-    deriving (Eq, Show, Functor, Generic, NFData)
+class
+    ( Ring c
+    , AdditiveGroup (poly c)
+    ) => UnivariateRingPolynomial poly c where
```

Comment:
As it does not link different instantiations of `poly` together (`poly c` and `poly d` for different `c` and `d` in a single function signature), I think it's better to set `poly` to be a `Type` rather than `Type -> Type`, like this:
```haskell
-- | A "main" type should go last to allow deriving.
-- A functional dependency wouldn't hurt.
class (Ring c, AdditiveGroup poly) => UnivariateRingPolynomial c poly | poly -> c where
  constant :: c -> poly -- actually can be just a `FromConstant`
  fromPoly :: poly -> V.Vector c
  ...
```